devtools/generate_supported.py

```diff
@@ -0,0 +1,253 @@
+    supported = {"kasa": {}, "tapo": {}}
+
+    _get_iot_supported(supported)
+    _get_smart_supported(supported)
+
+    readme_updated = _update_supported_file(
+        README_FILENAME, _supported_summary(supported), print_diffs
+    )
+    supported_updated = _update_supported_file(
+        SUPPORTED_FILENAME, _supported_detail(supported), print_diffs
+    )
+    if not readme_updated and not supported_updated:
+        print("Supported devices unchanged.")
+
+
+def _update_supported_file(filename, supported_text, print_diffs) -> bool:
+    with open(filename) as f:
+        contents = f.readlines()
+
+    start_index = end_index = None
+    for index, line in enumerate(contents):
+        if line == "<!--SUPPORTED_START-->\n":
+            start_index = index + 1
+        if line == "<!--SUPPORTED_END-->\n":
+            end_index = index
+
+    current_text = "".join(contents[start_index:end_index])
+    if current_text != supported_text:
+        print(
+            f"{filename} has been modified with updated "
+            + "supported devices, add file to commit."
+        )
+        if print_diffs:
+            print("##CURRENT##")
+            print(current_text)
+            print("##NEW##")
+            print(supported_text)
+
+        new_contents = contents[:start_index]
+        end_contents = contents[end_index:]
+        new_contents.append(supported_text)
+        new_contents.extend(end_contents)
+
+        with open(filename, "w") as f:
+            new_contents_text = "".join(new_contents)
+            f.write(new_contents_text)
+        return True
+    return False
+
+
+def _supported_summary(supported):
+    return _supported_text(
+        supported,
+        "### Supported $brand$auth devices\n\n$types\n",
+        "- **$type_**: $models\n",
+    )
+
+
+def _supported_detail(supported):
+    return _supported_text(
+        supported,
+        "## $brand devices\n\n$preamble\n\n$types\n",
+        "### $type_\n\n$models\n",
+        "- **$model**\n$versions",
+        "  - Hardware: $hw$region / Firmware: $fw$auth_flag\n",
+    )
+
+
+def _supported_text(
+    supported, brand_template, types_template, model_template="", version_template=""
+):
+    brandt = Template(brand_template)
+    typest = Template(types_template)
+    modelt = Template(model_template)
+    versst = Template(version_template)
+    brands = ""
+    version: SupportedVersion
+    for brand, types in supported.items():
+        preamble_text = (
+            "<sup>*</sup>&nbsp; Kasa model requires authentication"
+            if brand == "kasa"
+            else "All Tapo devices require authentication"
+        )
+        brand_text = brand.capitalize()
+        brand_auth = r"<sup>\*</sup>" if brand == "tapo" else ""
+        types_text = ""
+        for supported_type, models in sorted(
+            # Sort by device type order in the enum
+            types.items(),
+            key=lambda st: list(SupportedDeviceType).index(st[0]),
+        ):
```

Comment:
I think we could avoid doing this by introducing a manually maintained dict, it's not that the device types change that often. Maybe something like:
```
type_map = {
    "Plugs": [DeviceType.Plug],
    "Wall switches": [DeviceType.Dimmer, DeviceType.Switch],
    ...
}
```
and iterate over these and checking if the type is in the given list, or what do you think?
```

Reply:
The type map is at the top of the file but goes in the other direction